File: app/src/main/java/com/dvinc/workouttimer/di/modules/DataModule.java
/*
 * Copyright (c) 2018 by Denis Verentsov (decsent@Yandex.ru)
 * All rights reserved.
 */

package com.dvinc.workouttimer.di.modules;

import android.arch.persistence.room.Room;
import android.content.Context;

import com.dvinc.workouttimer.data.db.TrainingsDatabase;
import com.dvinc.workouttimer.data.repositories.laps.LapsRepository;
import com.dvinc.workouttimer.data.repositories.laps.LapsRepositoryImpl;
import com.dvinc.workouttimer.data.repositories.training.TrainingsRepository;
import com.dvinc.workouttimer.data.repositories.training.TrainingsRepositoryImpl;
import com.dvinc.workouttimer.di.qualifiers.IoScheduler;
import com.dvinc.workouttimer.di.qualifiers.UiScheduler;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import io.reactivex.Scheduler;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

@Module
public class DataModule {
    @Provides
    @Singleton
    TrainingsDatabase provideTrainingDataBase(Context context) {
        return Room.databaseBuilder(context, TrainingsDatabase.class, "trainingDatabase").build();
    }

    @Provides
    @Singleton
    TrainingsRepository provideTrainingsRepository(TrainingsDatabase trainingsDatabase) {
        return new TrainingsRepositoryImpl(trainingsDatabase);
    }

    @Provides
    @Singleton
    LapsRepository provideLapsRepository(TrainingsDatabase trainingsDatabase) {
        return new LapsRepositoryImpl(trainingsDatabase);
    }

    @Provides
    @IoScheduler
    Scheduler provideIoScheduler() {
        return Schedulers.io();
    }

    @Provides
    @UiScheduler
    Scheduler provideUiScheduler() {
        return AndroidSchedulers.mainThread();
    }
}

File: app/src/main/kotlin/com/dvinc/workouttimer/presentation/di/module/DataModule.kt
/*
 * Copyright (c) 2018 by Denis Verentsov (decsent@yandex.ru)
 * All rights reserved.
 */

package com.dvinc.workouttimer.presentation.di.module

import android.arch.persistence.room.Room
import android.content.Context
import com.dvinc.workouttimer.data.database.WorkoutDatabase
import com.dvinc.workouttimer.data.database.dao.ExerciseDao
import com.dvinc.workouttimer.data.database.dao.WorkoutDao
import dagger.Module
import dagger.Provides
import javax.inject.Singleton

@Module
class DataModule {

    companion object {
        private const val DATABASE_NAME = "workoutDb"
    }

    @Provides
    @Singleton
    fun provideDatabase(context: Context): WorkoutDatabase {
        return Room.databaseBuilder(context, WorkoutDatabase::class.java, DATABASE_NAME).build()
    }

    @Provides
    @Singleton
    fun provideWorkoutDao(database: WorkoutDatabase): WorkoutDao {
        return database.workoutDao()
    }

    @Provides
    @Singleton
    fun provideExerciseDao(database: WorkoutDatabase): ExerciseDao {
        return database.exerciseDao()
    }
}
